framework/src/utils/MooseMeshXYCuttingUtils.C

```diff
@@ -0,0 +1,1230 @@
+              : cross_elem->subdomain_id(),
+          tri_nodes[(node_finder + 1) % 3] == node_pairs_vec[i][node_3_finder > node_4_finder].first
+              ? cross_elem->subdomain_id()
+              : block_id_to_remove,
+          new_boundary_id);
+      mesh.delete_elem(cross_elem);
+    }
+  }
+  mesh.contract();
+
+  // Delete the block to remove
+  for (auto elem_it = mesh.active_subdomain_elements_begin(block_id_to_remove);
+       elem_it != mesh.active_subdomain_elements_end(block_id_to_remove);
+       elem_it++)
+    mesh.delete_elem(*elem_it);
+  mesh.contract();
+}
+
+void
+lineRemoverCutElem(ReplicatedMesh & mesh,
+                   const std::vector<Real> & cut_line_params,
+                   const dof_id_type tri_subdomain_id_shift,
+                   const SubdomainName tri_elem_subdomain_name_suffix,
+                   const subdomain_id_type block_id_to_remove,
+                   const boundary_id_type new_boundary_id,
+                   const bool improve_boundary_tri_elems)
+{
+  // Convert any quad elements crossed by the line into tri elements
+  quadToTriOnLine(mesh, cut_line_params, tri_subdomain_id_shift, tri_elem_subdomain_name_suffix);
+  // Then do the cutting for the preprocessed mesh that only contains tri elements crossed by the
+  // cut line
+  lineRemoverCutElemTri(mesh, cut_line_params, block_id_to_remove, new_boundary_id);
+
+  if (improve_boundary_tri_elems)
+    boundaryTriElemImprover(mesh, new_boundary_id);
+}
+
+void
+boundaryTriElemImprover(ReplicatedMesh & mesh, const boundary_id_type boundary_to_improve)
+{
+  if (!MooseMeshUtils::hasBoundaryID(mesh, boundary_to_improve))
+    mooseError(
+        "MooseMeshXYCuttingUtils::boundaryTriElemImprover(): The boundary_to_improve provided "
+        "does not exist in the given mesh.");
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  auto side_list = boundary_info.build_side_list();
+  // Here we would like to collect the following informaton for all the TRI3 elements on the
+  // boundary: Key: node id of the off-boundary node Value: a vector of tuples, each tuple contains
+  // the following information:
+  // 1. The element id of the element that is on the boundary to improve
+  // 2. the one node id of that element that is on the boundary to improve
+  // 3. the other node id of the element that is on the boundary to improve
+  std::map<dof_id_type, std::vector<std::tuple<dof_id_type, dof_id_type, dof_id_type>>>
+      tri3_elem_info;
+  for (const auto & side : side_list)
+  {
+    if (std::get<2>(side) == boundary_to_improve)
+    {
+      Elem * elem = mesh.elem_ptr(std::get<0>(side));
+      if (elem->type() == TRI3)
+      {
+        const auto key_node_id = elem->node_id((std::get<1>(side) + 2) % 3);
+        const auto value_elem_id = elem->id();
+        const auto value_node_id_1 = elem->node_id(std::get<1>(side));
+        const auto value_node_id_2 = elem->node_id((std::get<1>(side) + 1) % 3);
```

Comment:
what if the winding order is the other way and side 1 is node 1 and node 0

Reply:
It think this will be taken care of by the code later.